tests/unit/defrag.cpp

```diff
@@ -0,0 +1,398 @@
+#include <boost/test/unit_test.hpp>
+#include <boost/filesystem.hpp>
+
+#include "eblob_wrapper.h"
+
+#include "library/blob.h"
+#include "library/datasort.h"
+#include "library/list.h"
+
+#include "eblob/eblob.hpp"
+
+#include <vector>
+#include <utility>
+#include <random>
+#include <iostream>
+#include <functional>
+
+#include <sysexits.h>
+#include <err.h>
+
+
+void initialize_eblob_config(eblob_config &config) {
+	config.blob_size = 10 * (1ULL << 30); // 10Gib
+	config.defrag_timeout = 0; // we don't want to autodefrag
+	config.defrag_time = 0;
+	config.defrag_splay = 0;
+	config.blob_flags |= EBLOB_USE_VIEWS;
+}
+
+
+void fill_eblob(eblob_wrapper &wrapper,
+		std::vector<item_t> &shadow_elems,
+		item_generator &generator,
+		size_t total_records) {
+	for (size_t index = 0; index != total_records; ++index) {
+		shadow_elems.push_back(generator.generate_item(index));
+		BOOST_REQUIRE_EQUAL(wrapper.insert_item(shadow_elems.back()), 0);
+	}
+}
+
+
+void init_datasort_cfg(eblob_wrapper &wrapper, datasort_cfg &cfg) {
+	memset(&cfg, 0, sizeof(cfg));
+	cfg.b = wrapper.get();
+	cfg.log = cfg.b->cfg.log;
+}
+
+
+class iterator_private {
+public:
+	explicit iterator_private(std::vector<item_t> &items_)
+	: items(items_)
+	{
+	}
+
+	std::vector<item_t> &items;
+	size_t number_checked = 0;
+};
+
+
+int iterate_callback(struct eblob_disk_control *dc,
+		     struct eblob_ram_control *rctl __attribute_unused__,
+		     int fd,
+		     uint64_t data_offset,
+		     void *priv,
+		     void *thread_priv __attribute_unused__)
+{
+	// TODO: BOOST_REQUIRE_EQUAL used because of 54-th version of boost
+	// Maybe it will be more comfortable with BOOST_TEST and error messages in more modern boost version.
+	BOOST_REQUIRE(!(dc->flags & BLOB_DISK_CTL_REMOVE));  // removed dc occured
+
+	iterator_private &ipriv = *static_cast<iterator_private*>(priv);
+	auto &items = ipriv.items;
+	BOOST_REQUIRE(ipriv.number_checked < items.size()); // index out of range
+
+	auto &item = items[ipriv.number_checked];
+	BOOST_REQUIRE(!item.removed);  // item removed
+	BOOST_REQUIRE(!item.checked); //  item already checked
+	BOOST_REQUIRE(dc->data_size == item.value.size());  // sizes mismatch
+
+	std::vector<char> data(dc->data_size);
+	int ret = __eblob_read_ll(fd, data.data(), dc->data_size, data_offset);
+	BOOST_REQUIRE(ret == 0);  // can't read data
+	BOOST_REQUIRE(data == item.value);  // contet of value differ
+
+	item.checked = true;
+	++ipriv.number_checked;
+	return 0;
+}
+
+
+int datasort(eblob_wrapper &wrapper, const std::set<size_t> &indexes) {
+	size_t number_bases = indexes.size();
+	assert(!indexes.empty());
+	assert(*indices.rbegin() < number_bases);
+
+	datasort_cfg dcfg;
+	memset(&dcfg, 0, sizeof(datasort_cfg));
+	std::vector<eblob_base_ctl *> bctls(number_bases);
+	eblob_base_ctl *bctl;
+	size_t index = 0;
+	size_t loop_index = 0;
+	list_for_each_entry(bctl, &wrapper.get()->bases, base_entry) {
+		if (indexes.count(loop_index)) {
+			bctls[index] = bctl;
```

Comment:
How about
```suggestion
			bctls.emplace_back(bctl);
```

and use `bctls.size()` instead of `index`?

Reply:
done